Update directory-structure.md: clarify service, service group, uniform versioning.

konrad-jamrozik opened the pull request:
@JeffreyRichter this PR is a follow-up based on our conversation. Please have a look at the [directory-structure.md diff](https://github.com/Azure/azure-rest-api-specs/pull/28929/files#diff-5b72db59365e9ce8cfc776cd9a1437ebba24e85e4dada20b5a7835aa39e97e22) and let me know what you think. I left some TODOs with open questions.

Some of information added could perhaps be better organized and even moved to a different article, but for now let's focus on having the facts right and reorganize the doc later.

This PR is a follow-up to:

- https://github.com/Azure/azure-rest-api-specs/pull/28606

Related:
- https://github.com/Azure/azure-sdk-tools/issues/8028
- https://github.com/Azure/azure-sdk-tools/issues/7628
- https://github.com/Azure/azure-rest-api-specs/issues/28873#issuecomment-2089174189
- https://github.com/Azure/azure-sdk-tools/issues/8203

Commit:
ongoing

Commit:
add some TODOs

konrad-jamrozik commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
+  version itself must be `-preview`, it can depend only **only** `-preview` versions, and the generated SDK
+  and published docs must denote they are in `preview` too.
+
+TODO: 
```
@JeffreyRichter first set of open questions I have.

konrad-jamrozik commented on documentation/directory-structure.md:
```diff
@@ -184,6 +242,19 @@ In case of big Azure teams, their `specification/<azureTeam>` hosts multiple ser
+
+Furthermore the `README.md` files are no longer located in the  `resource-manager` or `data-plane` folders, but instead in `<service>` folders.
+
+TODO:
```
@JeffreyRichter second set of open questions I have.

konrad-jamrozik commented on documentation/directory-structure.md:
```diff
@@ -211,6 +282,11 @@ This means that each service in the service group must obey the same versioning
 
+### Migrating from singular service to service group
+
+TODO: how do we do it? This is the case of:
```
@JeffreyRichter third set of open questions I have.

mikekistler commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
+In practice, this means, for an example service with version `2024-03-25`:
+
+- All the OpenAPI specification `.json` files the service is composed of must have the same version of `2024-03-25`.
+- All the shared specifications (i.e. `common-types`) the service depends on must have the same version of `2024-03-25`.
```
Common types typically do not use date-based versioning. But we can say that the versions should all match.
```suggestion
- All the shared specifications (i.e. `common-types`) the service depends on must have the same version.
```

mikekistler commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
+  In addition, a new SDK must be generated from the service, and new documentation published, both tagged with
+  version `2024-04-17`.
+- As a consequence of the above, if given service version uses `-preview` version anywhere, then the service
+  version itself must be `-preview`, it can depend only **only** `-preview` versions, and the generated SDK
```
```suggestion
  version itself must be `-preview`, it can depend **only** on `-preview` versions, and the generated SDK
```

mikekistler commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
+  and published docs must denote they are in `preview` too.
+
+TODO: 
+- These rules seem too strict. What if cadence of releasing `common-types` does not align with cadence of releasing
+  given service? Same for resource types? 
+  - For example, service got released on `2024-01-01`, then new `common-types` got released on `2024-02-01` and on
+    `2024-03-01` new service version is to be released? Can a `2024-03-01` service version depend on `2024-02-01` `common-types`?
+  - If a service version can depend on `common-types` of a different version, can the generated SDK also depend
+    on different `.json` versions?
+    See e.g. [this bad setup of compute rm 2024-03-01](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/readme.md#tag-package-2024-03-01).
+    Could we make package `2024-03-01` depend on, say, `common-types` of `2024-02-01` and all other `.json` files of, say, `2024-02-15` ? 
```
I think most of these concerns are moot if you accept the change I proposed above.

mikekistler commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
+    Could we make package `2024-03-01` depend on, say, `common-types` of `2024-02-01` and all other `.json` files of, say, `2024-02-15` ? 
+- What is the difference between resource types and OpenAPI specification `.json` files?
+  Can one deduce the resource type version by looking at the source OpenAPI `.json` ?
+- What does it mean for OpenAPI specification `.json` to "have a version"? Are we talking about it being in appropriate
```
I had this same question when reading the text above. I think what this means -- and we should be specific about it -- is that the file resides in a `/prevew/<version>` or `/stable/<version>` folder.

mikekistler reviewed (commented):
Left a few small edits.

Commit:
Apply suggestions from code review

Co-authored-by: Mike Kistler <mikekistler@microsoft.com>

Commit:
elaborate

Commit:
clarify

Commit:
clarify

Commit:
clarify

Commit:
clarify

Commit:
conciseness

mikekistler reviewed (approved):
Looks good! 👍

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
 
+## Key concepts
+
+The directory structure is a reflection of few key concepts.
```
```suggestion
The directory structure is a reflection of a few key concepts.
```

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+
+A `service` is a set of operation endpoints (typically HTTP REST API endpoints) that **version uniformly**.
+
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
```
```suggestion
A `service group` is a set of services that share a common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
```

JeffreyRichter replied:
I think it's important to say that "services within a group can version independently of each other." Maybe put this in the "Uniform versioning" section? This independent versioning (with different docs, SDKs, ship timeframes, etc.) is the main motivation for having groups in the first place.

markcowl replied:
Does this mean that, within a workgroup RP namespace, api-versions on the wire can be non-uniforrm across the resource namespace,  meaning that if service A uses api-version 2024-01-01 and service B uses api-version 2024-02-01, that service A's endpoints may not be available under api-version 2024-02-01?

rkmanda replied:
I think whoever initially came up with the name 'service group' could have clearly picked a better name. A service group as it stands today is a collection of resource types. That collection must version uniformly. A service is either a collection of resource types or a collection of service groups. We could possibly brainstorm on better names.

rkmanda replied:
Heres the hierarchy I am thinking of - 

Service - A collection of resource types or resource type groups. A service must be homogenous, so it must not have both resource types and resource type groups at the same level. 

Resource type groups - A collection of resource types. All resource types under a resource type group must version uniformly.

Resource type - An abstraction for a server side entity that is addressable through a url and operated on using HTTP methods.

konrad-jamrozik replied:
@rkmanda

Re `service group`. Consider these two distinct `services`:
- https://learn.microsoft.com/en-us/azure/aks/
- https://learn.microsoft.com/en-us/azure/kubernetes-fleet/

Both of them belong to the same `service group`. They share the same dir:
[`specification/containerservice/resource-manager/Microsoft.ContainerService`](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/containerservice/resource-manager/Microsoft.ContainerService)

Seems to me this is a missing element from the hierarchy you proposed. 

Second thing: what is an effective distinction between a service that is a "collection of resource types" vs a "collection of resource type groups" ? How is this different from the point of view of our users? Is the end-user documentation organized differently, or the SDKs? Can you point me to an example? Here is a [resource provider list](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers#match-resource-provider-to-service).

Because there is a distinction between `serviced and `service group` from the point of view of our users: 
- `service group` doesn't have to version uniformly, but the services it is composed of must
- Each `service` has its own corresponding SDK and documentation portal, but `service group` does not.

In your proposal, if the service is a collection of "resource type groups", then it would mean these **do not** have to version uniformly. Each `resource type group` has to version uniformly, but service is composed of multiple such groups. We cannot allow that. Entire service has to version uniformly, so a service **can not** be composed of **multiple** `resource type groups`.

Finally, you wrote:

>  A service is either a collection of resource types or a collection of service groups. 

This doesn't make sense to me, because I don't see how a service can be a collection of service groups. A service group is a collection of services, not the other way around (and with two levels of aggregation: the group itself is a collection, so a collection of service groups would be a collection of collection of services)

JeffreyRichter replied:
@rkmanda Do we need the concept of a "resource type group"? Can we just say that a "service" is set of "resource types" that version uniformly? In other words, I think a "resource type group" is synonymous with a "service" and therefore we do not need 2 terms to represent a single abstraction.

Do you disagree with this thinking? If so, then I need to have the concept of "resource type group" explained to me.

JeffreyRichter replied:
> I think whoever initially came up with the name 'service group' could have clearly picked a better name. A service group as it stands today is a collection of resource types. That collection must version uniformly. A service is either a collection of resource types or a collection of service groups. We could possibly brainstorm on better names.

@rkmanda I'm trying very hard to have the term "service" be what customers understand as the thing that versions, has docs/version, has an SDK/version, has a preview or stable version, etc. And then, we want to use the term "service" both within Microsoft and outside of Microsoft so we all use common terminology.

I'm trying to NOT burden customers with concepts of resource types, service groups, etc. These would be MS-internal only concepts used to give names to abstractions that WE talk abut but that customers have no need to talk about.

JeffreyRichter replied:
> Does this mean that, within a workgroup RP namespace, api-versions on the wire can be non-uniforrm, meaning that if service A uses api-version 2024-01-01 and service B uses api-version 2024-02-01, that service A's endpoints may not be available under api-version 2024-02-01?

@markcowl Yes, there may be a "service-A" whose endpoints all use 2024-01-01 and another "service-B" whose endpoints all use 2024-02-1. But, I'm not sure what your use of "under" means in your question since endpoints are all flat.

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
+
+In case of ARM, a `service` consists of multiple `resource types`.
+A `resource type` can be derived from the URL of HTTP API operations that pertain to it.
```
While this is correct and good for your understanding, I don't know if it worth mentioning in this doc. 
If you want to keep it, you might also want to add that "Within a service, a resource types must share the same api-version".

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+   one service version.
+1. The service version must be always represented in its entirety; any SDK or documentation referring to the service
+   must encompass all of it.
+1. The `common-types` OpenAPI definition shared across multiple services can version independently of the service.
```
I'm not sure what you mean by this. What we need to say is something like this: 
"Any change to `common-types` requires all services depending on it to also version.

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+1. The service version must be always represented in its entirety; any SDK or documentation referring to the service
+   must encompass all of it.
+1. The `common-types` OpenAPI definition shared across multiple services can version independently of the service.
+   However, the rule that API version is immutable still must be observed.
```
Previous api-versions will remain immutable because new api-versions must be created when versioning `common-types`.

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+The **uniform versioning** has several implications and implementation decisions supporting it:
+
+- The service is effectively defined as a series of consecutive API versions.
+- Each API version is represented by a pair of folders representing its lifecycle stage and release date, 
```
It's not actually a release date. It is just a date. It _could_ be a release date, or it could be the date when the new version started being implemented. The only requirement is that each date me later in time than the previous date - the date has no actual meaning.

JeffreyRichter replied:
One other thing: it is illegal to have both stable and preview versions with the same date. For example, 2024-05-08-preview and 2024-05-08 should NEVER happen. The reason is because it is impossible for the customer to tell if the stable came after the preview or vice verse.  Some teams might try to go from preview to stable by simply truncating the `-preview` suffix; this is NOT acceptable. Instead, the team would have to pick a later date for the stable which could be just 1 day later: 2024-05-09

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
+  of must have the same version of `2024-04-17` and the `info.version` property must say `2024-04-17`.
+  In addition, a new SDK must be generated from the service, and new documentation published, both tagged with
+  service version `2024-04-17`.
+- All of these rules apply both for OpenAPI specs emitted from TypeSpec as well as for hand-written.
```
Yes, all of the above is correct

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -211,6 +286,18 @@ This means that each service in the service group must obey the same versioning
+  inside a directory denoting the service name (i.e. `/<service>/`), and its README.md file will also be moved there. 
+  As this PR is moving the service, it means the PR diff will include deletions of the singular service README.md and
+  API version folders, and it will include addition of the new README.md and API version folder,
+  both inside the service folder.
```
What you describe here is one option and whether to allow this or not is really up to your team and how hard it is for you and your tooling to deal with it. 
Another option is that the 1st service of the 'service group' must stay as it exactly is - this allow diffs to work well.
Then any NEW services go in a subdir with its readme.md & stable/preview folders. 
The end result is that the 1st service in the group is in the parent and ALL OTHER services in the group are in their respective subdirectories.

Your team can pick ONE of these or allow either of these depending on what works best for your team. If you allow moving the 1st service to a subdir, then we would still need all the diff tooling to work correctly and I think this would be hard for your team to accomplish.

rkmanda reviewed (changes requested):
:clock1:

Commit:
Update docs

konrad-jamrozik commented:
@JeffreyRichter @rkmanda can you have another look? The additional constraints I placed on the `service` folder location have been agreed to by @rkmanda. Similarly, while the `service group` terminology has been previously confused, we will work towards sorting out the way I proposed.

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,140 +30,102 @@ You may be also interested in following:
-## AutoRest configuration `README.md` files
+- `<azureTeam>/<typeSpecSrc>` (multiple `<TypeSpecSrc>` folders are allowed)
+- `<azureTeam>/resource-manager/<RPNS>` (exactly one `<RPNS>` folder is allowed)
+- `<azureTeam>/data-plane/<namespace>` (multiple `<namespace>` folders are allowed)
```
What is a namespace for data-plane? I don't think we have this concept. If we do, then it needs to be in the glossary.

I did some scanning of the repo and I think the "namespace" concept was originally for the .NET SDK namespace which doesn't map to all the languages we support today. The .NET namespace is al cilent-side concept and not a service-side concept (like RPNS). If I'm right, then we need to clean this up too.

konrad-jamrozik replied:
@JeffreyRichter I proposed to call it `<namespace>` based on all the `data-plane` children folder  names I observed. I provide a full list at the end of this comment.

All the guidance we had so far said that `data-plane` child folders can be arbitrary.

If you don't like the idea of naming it `<namespace>` do you have another proposal? `<serviceGroupName>` ? In case of ARM `resource-manager/<RPNS>` corresponds exactly 1 to 1 to `resource-manager/"service group name"`

Here is the full list of data-plane child folder names, deduplicated:

```
/data-plane/AnomalyDetector
/data-plane/ApiCenter.DataApi
/data-plane/AsyncBatchManagement
/data-plane/AutoSuggest
/data-plane/Azure.AI.ChatProtocol
/data-plane/Azure.AI.DocumentTranslation
/data-plane/Azure.AI.TextTranslation
/data-plane/Azure.Analytics.Purview.Account
/data-plane/Azure.Analytics.Purview.Catalog
/data-plane/Azure.Analytics.Purview.DataMap
/data-plane/Azure.Analytics.Purview.DevopsPolicies
/data-plane/Azure.Analytics.Purview.MetadataPolicies
/data-plane/Azure.Analytics.Purview.PDS
/data-plane/Azure.Analytics.Purview.Scanning
/data-plane/Azure.Analytics.Purview.SelfServicePolicies
/data-plane/Azure.Analytics.Purview.Share
/data-plane/Azure.Analytics.Purview.Workflow
/data-plane/Azure.Batch
/data-plane/Azure.ContainerRegistry
/data-plane/Azure.Contoso.WidgetManager
/data-plane/Azure.Developer.Signing
/data-plane/Azure.ProgrammableConnectivity
/data-plane/Azure.Search
/data-plane/Azure.Storage.Files.DataLake
/data-plane/AzureOpenAI
/data-plane/CallAutomation
/data-plane/CallingServer
/data-plane/Chat
/data-plane/Common
/data-plane/ComputerVision
/data-plane/ContentModerator
/data-plane/ContentSafety
/data-plane/Creator
/data-plane/CustomImageSearch
/data-plane/CustomVision
/data-plane/CustomWebSearch
/data-plane/DEM
/data-plane/Data
/data-plane/DataRegistry
/data-plane/DocumentIntelligence
/data-plane/Email
/data-plane/EntitySearch
/data-plane/Face
/data-plane/FormRecognizer
/data-plane/Geolocation
/data-plane/HealthInsights
/data-plane/Identity
/data-plane/ImageAnalysis
/data-plane/ImageSearch
/data-plane/JobRouter
/data-plane/LUIS
/data-plane/Language
/data-plane/LiveMetrics
/data-plane/LocalSearch
/data-plane/MediaComposition
/data-plane/Messages
/data-plane/MetricsAdvisor
/data-plane/Microsoft.AVS
/data-plane/Microsoft.AgFoodPlatform
/data-plane/Microsoft.ApiCenter
/data-plane/Microsoft.ApiManagement
/data-plane/Microsoft.AppConfiguration
/data-plane/Microsoft.Attestation
/data-plane/Microsoft.Azure.Search.Data
/data-plane/Microsoft.Azure.Search.Service
/data-plane/Microsoft.Batch
/data-plane/Microsoft.BlobStorage
/data-plane/Microsoft.Cache
/data-plane/Microsoft.CodeTransparency
/data-plane/Microsoft.Communication
/data-plane/Microsoft.ConfidentialLedger
/data-plane/Microsoft.ContainerRegistry
/data-plane/Microsoft.ContainerService
/data-plane/Microsoft.DataBox
/data-plane/Microsoft.DataLakeAnalytics
/data-plane/Microsoft.DataLakeStore
/data-plane/Microsoft.DevCenter
/data-plane/Microsoft.DeviceUpdate
/data-plane/Microsoft.Devices
/data-plane/Microsoft.DigitalTwins
/data-plane/Microsoft.Easm
/data-plane/Microsoft.EventGrid
/data-plane/Microsoft.EventHub
/data-plane/Microsoft.FileStorage
/data-plane/Microsoft.GraphRbac
/data-plane/Microsoft.HDInsight
/data-plane/Microsoft.HealthcareApis
/data-plane/Microsoft.Insights
/data-plane/Microsoft.InstanceMetadataService
/data-plane/Microsoft.IoTCentral
/data-plane/Microsoft.KeyVault
/data-plane/Microsoft.LoadTestService
/data-plane/Microsoft.MachineLearningServices
/data-plane/Microsoft.ManagedCcf
/data-plane/Microsoft.Maps
/data-plane/Microsoft.Marketplace
/data-plane/Microsoft.Media
/data-plane/Microsoft.MixedReality
/data-plane/Microsoft.OperationalInsights
/data-plane/Microsoft.PlaywrightTesting.AuthManager
/data-plane/Microsoft.PolicyInsights
/data-plane/Microsoft.Quantum
/data-plane/Microsoft.QueueStorage
/data-plane/Microsoft.ResourceNotifications
/data-plane/Microsoft.Resources
/data-plane/Microsoft.SchemaRegistry
/data-plane/Microsoft.SecurityInsights
/data-plane/Microsoft.ServiceBus
/data-plane/Microsoft.ServiceFabric
/data-plane/Microsoft.SignalRService
/data-plane/Microsoft.Storage
/data-plane/Microsoft.StorageDataLake
/data-plane/Microsoft.Synapse
/data-plane/Microsoft.Tables
/data-plane/Microsoft.TimeSeriesInsights
/data-plane/Microsoft.VoiceServices
/data-plane/Microsoft.Web
/data-plane/Monitor.Exporters
/data-plane/NetworkTraversal
/data-plane/NewsSearch
/data-plane/OpenAI.Assistants
/data-plane/OperatorConnect
/data-plane/Personalizer
/data-plane/PhoneNumbers
/data-plane/Products
/data-plane/QnAMaker
/data-plane/Render
/data-plane/Reviews
/data-plane/Rooms
/data-plane/Route
/data-plane/Search
/data-plane/SipRouting
/data-plane/Sms
/data-plane/Spatial
/data-plane/SpeakerRecognition
/data-plane/Speech
/data-plane/SpellCheck
/data-plane/TextAnalytics
/data-plane/Timezone
/data-plane/Traffic
/data-plane/TranslatorText
/data-plane/UnifiedVision
/data-plane/VideoAnalyzer.Edge
/data-plane/VideoSearch
/data-plane/VisualSearch
/data-plane/Weather
/data-plane/WebPubSub
/data-plane/WebSearch
/data-plane/ingestion
/data-plane/preview
/data-plane/v1
```

Note the `/data-plane/preview`. There is one instance of it, in [monitor](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/monitor/data-plane/preview)

JeffreyRichter replied:
I can't tell from the list above is a single "service team" has 2+ data-plane subdirs with README.md files. This would be useful to know to see if we have to support "service groups" for data-plane. Maybe these subdirs are the equivalent of "services" within a "service group" for data-plane.

Does anything actually use this subdir name for anything? If anything did, it would be an SDK and I do not think they use it. And, if they did, it is probably only the .NET SDK.
Can we find out if any tool actually uses this subdir name for anything?

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,140 +30,102 @@ You may be also interested in following:
-Each `README.md` describes a single `service` and is used as an SDK package and documentation for each version of the service.
-Inside the `README.md` file there are lists of paths to OpenAPI spec `.json` files making up given service version.
+The content of these folders is used as input to emit OpenAPI specifications
+placed within `<azureTeam>/resource-manager/<RPNS>` and `<azureTeam>/data-plane/<namespace>` folders.
```
namespace?

konrad-jamrozik replied:
Let's discuss here:
https://github.com/Azure/azure-rest-api-specs/pull/28929#discussion_r1625472887

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -30,140 +30,102 @@ You may be also interested in following:
+shared across API versions. E.g. [`Microsoft.Compute/common-types`].
 
-# ===== data-plane
+### `data-plane/<namespace>` folders
```
namespace?
I won't point out future uses of namespace but I still question them.

konrad-jamrozik replied:
Let's discuss here:
https://github.com/Azure/azure-rest-api-specs/pull/28929#discussion_r1625472887

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -173,75 +135,65 @@ For another example, see [`specification/eventgrid`].
+  it contains mixture of both the deprecated structure (for the first service) as well as the correct structure (for the 2nd and following services).
 
-Using our example, note the most important directory structure difference of a `service group` scenario as compared to singular `service`:
+This legacy structure is not allowed going forward.
```
Are you sure you don't want to allow this?
I think it would be OK and simplifies processes.

konrad-jamrozik replied:
@JeffreyRichter Yes. 

- First we agree to do that
- Then we will document it (I will merge this PR)
- Then we will diffuse the knowledge to all PR reviewers, (ARM, SDK gen, etc.)
- Then we will enforce it in tooling. We have some work todo this properly (details [in this comment](https://github.com/Azure/azure-rest-api-specs/issues/29065#issuecomment-2104104232)) but I think we could quickly deploy some simple check validating just that.

@mikeharder does that sound good to you too?

konrad-jamrozik replied:
I synced with @mikeharder and we agreed. Details here:

- https://github.com/Azure/azure-rest-api-specs/issues/29065#issuecomment-2148357551

konrad-jamrozik replied:
We chatted in person. Meeting notes here:
- https://github.com/Azure/azure-rest-api-specs/pull/28929#issuecomment-2151095434

JeffreyRichter commented on documentation/directory-structure.md:
```diff
@@ -173,75 +135,65 @@ For another example, see [`specification/eventgrid`].
-  instead of directly containing `stable` and `preview` folders.
-- Each of the `aks` and `fleet` subfolders have their own `README.md` file, instead of there being one `README.md` in the `resource-manager` folder.
-  As a result, the SDKs of these services are separate.
+## Migrating from a singular service to service group in a legacy directory structure
```
I'm sure what you mean by "legacy". Do we NEVER want teams to use the old structure that has everything at the root and ALWAYS create subdirs even if the RPNS has just 1 service in its service group?
If we NEVER want them to do this, then "legacy" is an OK word to use otherwise it is not legacy but just an "alternative" dir structure.

konrad-jamrozik replied:
Yes. Let us discuss here:
https://github.com/Azure/azure-rest-api-specs/pull/28929#discussion_r1619337954

Commit:
Merge branch 'main' into users/kojamroz/dir_struct_2

Commit:
Merge branch 'main' into users/kojamroz/dir_struct_2

Commit:
Soften guidance

konrad-jamrozik commented:
Today @JeffreyRichter, @MikeHarder, @MSEvanhi, @jhendrixMSFT and I met to discuss this work. Evan will prepare a proposal for a bigger overhaul of the specification directory structure.

@JeffreyRichter, given we are delaying the "ultimate desired state" of the specs' directory structure to the proposal, it will take some time until it materializes. Because of this, I updated this pull request under the assumption that we no longer **enforce** a better layout to current directory structure, but instead we **recommend** it, to help people do the right thing and provide updated guidance in the short-term.

I also changed the data-plane `<namespace>` concept to `<groupingDir>` to better reflect the current state of things.

In the short to medium term, we will work on automation enforcing the uniform versioning, unless you see a reason to not do that until we have the new directory structure new proposal deployed.

I would kindly ask you to review the PR and approve, unless you see blockers. Thank you!

Commit:
Merge branch 'main' into users/kojamroz/dir_struct_2

MaryGao commented on documentation/uniform-versioning.md:
```diff
@@ -0,0 +1,95 @@
+
+**Uniform versioning** prescribes the following
+
+1. Any deployed service operation endpoint must belong to an API version. An API version once deployed is immutable.
```
@konrad-jamrozik I have two questions regarding this:
- Considering we have no issues in service side but issues in defined spec in one version, can we directly update spec to fix this spec issue?
- Considering we have issues in service side and relevant spec in one version, can we directly fix the issue in service side and update the spec in the same api version?

konrad-jamrozik replied:
@MaryGao:

Re

> Imagine we have no issues in service side but issues in defined spec in v1, can we directly update spec to fix this spec issue in v1?

Yes, you can. If there is a breaking change caused by the spec update to match the service behavior, you can approve it with `BreakingChange-Approved-BugFix` or `Versioning-Approved-BugFix`. Details:

- https://aka.ms/azsdk/pr-brch-deep#breakingchange-approved-bugfix (`BreakingChange-Approved-BugFix`)
- https://aka.ms/azsdk/pr-brch-deep#versioning-approved-bugfix (`Versioning-Approved-BugFix`)

Re 

> Imagine we have issues in service side and relevant spec in v2, can we directly fix the issue in service side and update the spec in the same api version v2?

I believe the answer is: no, you can not. But if this was a `preview` API version, you can introduce a new API version without going through the formal "breaking change notification" process (the green "ignore" box in [this diagram](https://eng.ms/docs/products/azure-developer-experience/design/specs-pr-guides/pr-brch-deep#diagram-explaining-breaking-changes-and-versioning-issues)). However, I am not 100% sure about this guidance - depending of the exact nature of the bug the Breaking Change Board may possibly allow you to do it. Let me defer to @mikekistler and @JeffreyRichter to clarify the guidance here.

mikekistler replied:
On the second point, if V2 is deployed and available to customers then its behavior should not change. If the behavior is "wrong" and needs to be fixed that should be done in a new api-version.

MaryGao commented on documentation/uniform-versioning.md:
```diff
@@ -0,0 +1,95 @@
+- Each API version is represented by a pair of folders representing its lifecycle stage and a date, 
+  like `stable/2024-03-05` or `preview/2024-05-15-preview`.
+- Each API version date must be later than the previous date.
+- `stable` and `preview` API versions cannot have the same date. This would prevent API users from knowing
```
Do we require the newly added version(no matter preview or stable) would be always latter than previous one?

Sometimes service team may have two preview versions e.g `preview/2023-05-15-preview` and `preview/2024-05-15-preview` but they want to stablelize the first one, we may have version `stable/2023-05-16`. Do we allow this?

konrad-jamrozik replied:
@MaryGao having more than 1 active preview is forbidden. In your case the `preview/2023-05-15-preview` should have been decommissioned and any new `stable` must be a successor to `preview/2024-05-15-preview`. The earliest date it can have is `stable/2024-05-16`.

See also:

[FALSE: My service supports preview features in API versions that came before our current GA](https://eng.ms/docs/products/azure-developer-experience/design/specs-pr-guides/pr-brch-deep#false-my-service-supports-preview-features-in-api-versions-that-came-before-our-current-ga)

@mikekistler and @JeffreyRichter can you double check my guidance here?

mikekistler replied:
I concur with this guidance.
